### PR TITLE
fix #2571

### LINKED ISSUE #2571: Some hexes can incorrectly act as impassable
Reported in the forums:
https://forums.wesnoth.org/viewtopic.php?f=4&t=47782
https://forums.wesnoth.org/viewtopic.php?p=624042#p624042

### PR DESCRIPTION
**Pull request by gfgtdf**

### CODE REVIEW

File: src/terrain/type_data.cpp [hunk 1]
```diff
@@ -151,24 +149,29 @@ std::string terrain_type_data::get_underlying_terrain_string(const t_translation
...
 		terrain_type new_terrain(base_iter->second, overlay_iter->second);
 		terrainList_.push_back(new_terrain.number());
-		tcodeToTerrain_.emplace(new_terrain.number(), new_terrain);
-		return true;
+		return tcodeToTerrain_.emplace(new_terrain.number(), std::move(new_terrain)).first;
 	}
-	return true; // Terrain already exists, nothing to do
+}
+
+bool terrain_type_data::try_merge_terrains(const t_translation::terrain_code & terrain)
```
Comment (jyrkive) [inline src/terrain/type_data.cpp]: There shouldn't be a space before the ampersand.

File: src/terrain/type_data.cpp [hunk 2]
```diff
@@ -65,6 +65,9 @@ const t_translation::ter_list& terrain_type_data::underlying_mvt_terrain(const t
 	auto i = find_or_merge(terrain);
 
 	if(i == tcodeToTerrain_.end()) {
+		// TODO: At least in some cases(for exampel when this is called form lua) it
```
Comment (jyrkive) [inline src/terrain/type_data.cpp]: - space before `(`
- example
- from

File: src/terrain/type_data.cpp [hunk 3]
```diff
@@ -65,6 +65,9 @@ const t_translation::ter_list& terrain_type_data::underlying_mvt_terrain(const t
...
+		// seems to make more sense to throw an excaption here, samge goes for get_terrain_info
```
Comment (jyrkive) [inline src/terrain/type_data.cpp]: - exception
- same

File: src/terrain/type_data.cpp (hunk 2 above)
Comment (gfgtdf) [reply src/terrain/type_data.cpp]: will fix

File: src/terrain/type_data.cpp (hunk 1 above)
Comment (gfgtdf) [reply src/terrain/type_data.cpp]: didn't write this code.. will probably still fix.

**Pull request merged**